src/lib.rs

```diff
@@ -1,6 +1,99 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//! Manages system resources that can be allocated to VMs and their devices.
+
 #![deny(missing_docs)]
-//! Dummy crate needs high-level documentation.
-/// Dummy public function needs documentation.
-pub fn it_works() {
-    assert!(true);
+
+mod id_allocator;
+
+pub use crate::id_allocator::IdAllocator;
+use std::fmt;
+use std::result;
+
+/// Error type for IdAllocator usage.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// All ids from the range specified are exhausted.
+    Overflow,
+    /// An id that is not part of the specified range was requested to be released.
+    OutOfRange(u32),
+    /// An id that was already released was requested to be released.
+    AlreadyReleased(u32),
+    /// An id  that was never allocated was requested to be released.
+    NeverAllocated(u32),
+    /// There are no more IDs available in the manage range
+    ResourceExhausted,
+    /// The range to manage is invalid.
+    InvalidRange(u32, u32),
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
```

Comment:
I might be wrong here, but I think that there's no need to define the `source()` method. The default implementation  that returns `None` is more appropriate, because the error itself should not be its own source. The source would be appropriate if, say, there was an I/O error leading to this one.

Reply:
There's no need to manually re-implement the default implementation of [source](https://doc.rust-lang.org/std/error/trait.Error.html#method.source). The empty impl is sufficient:
```rust
impl std::error::Error for Error {}
```